### PR TITLE
Add test to show snapshot doesn't block shard close/fail

### PR DESCRIPTION
**Pull request by DaveCTurner**
Relates ES-9635

### CODE REVIEW

File: server/src/internalClusterTest/java/org/elasticsearch/snapshots/SharedClusterSnapshotRestoreIT.java [hunk 1]
```diff
@@ -1405,6 +1409,84 @@ public void testCloseOrDeleteIndexDuringSnapshot() throws Exception {
...
+        assertFalse(snapshotFuture.isDone());
+
+        try {
+            if (randomBoolean()) {
+                logger.info("--> closing index [{}]", indexName);
+                safeGet(indicesAdmin().prepareClose(indexName).execute());
+            } else {
+                logger.info("--> failing index [{}] to trigger recovery", indexName);
+                for (IndexService indexService : internalCluster().getInstance(IndicesService.class, blockingNode)) {
+                    if (indexService.index().getName().equals(indexName) != false) {
```
Comment (henningandersen) [inline server/src/internalClusterTest/java/org/elasticsearch/snapshots/SharedClusterSnapshotRestoreIT.java:1467]: You had me for a while here :slightly_smiling_face: , can we simplify this to:

```suggestion
                    if (indexService.index().getName().equals(indexName)) {
```

File: server/src/internalClusterTest/java/org/elasticsearch/snapshots/SharedClusterSnapshotRestoreIT.java [hunk 2]
```diff
@@ -1405,6 +1409,84 @@ public void testCloseOrDeleteIndexDuringSnapshot() throws Exception {
...
+        }
+        assertFalse(snapshotFuture.isDone());
+
+        try {
+            if (randomBoolean()) {
+                logger.info("--> closing index [{}]", indexName);
+                safeGet(indicesAdmin().prepareClose(indexName).execute());
+            } else {
+                logger.info("--> failing index [{}] to trigger recovery", indexName);
+                for (IndexService indexService : internalCluster().getInstance(IndicesService.class, blockingNode)) {
```
Comment (henningandersen) [inline server/src/internalClusterTest/java/org/elasticsearch/snapshots/SharedClusterSnapshotRestoreIT.java:1466]: Can we verify in the test that we get to the failshard part?

File: server/src/internalClusterTest/java/org/elasticsearch/snapshots/SharedClusterSnapshotRestoreIT.java [hunk 3]
```diff
@@ -1405,6 +1409,84 @@ public void testCloseOrDeleteIndexDuringSnapshot() throws Exception {
...
+        flushAndRefresh(indexName);
+
+        safeAwait(barrier);
+
+        final var snapshotName = randomIdentifier();
+        ActionFuture<CreateSnapshotResponse> snapshotFuture = clusterAdmin().prepareCreateSnapshot(
+            TEST_REQUEST_TIMEOUT,
+            repoName,
+            snapshotName
+        ).setIndices(indexName).setWaitForCompletion(true).setPartial(true).execute();
```
Comment (henningandersen) [inline server/src/internalClusterTest/java/org/elasticsearch/snapshots/SharedClusterSnapshotRestoreIT.java:1447]: I would think that it is not important to this test whether the snapshot it partial or not - only to what the final `safeGet` does (throw or not). Can we randomize it?

**Review by henningandersen — APPROVED** [review]
LGTM.

File: server/src/internalClusterTest/java/org/elasticsearch/snapshots/SharedClusterSnapshotRestoreIT.java (hunk 1 above)
Comment (DaveCTurner) [reply server/src/internalClusterTest/java/org/elasticsearch/snapshots/SharedClusterSnapshotRestoreIT.java]: Ah sorry that was an automatic refactoring by IntelliJ that I didn't notice needed cleanup

File: server/src/internalClusterTest/java/org/elasticsearch/snapshots/SharedClusterSnapshotRestoreIT.java (hunk 2 above)
Comment (DaveCTurner) [reply server/src/internalClusterTest/java/org/elasticsearch/snapshots/SharedClusterSnapshotRestoreIT.java:1469]: Sure see c222218f57bf902cbb2e8fe9d947c2bc619d8276.

File: server/src/internalClusterTest/java/org/elasticsearch/snapshots/SharedClusterSnapshotRestoreIT.java (hunk 3 above)
Comment (DaveCTurner) [reply server/src/internalClusterTest/java/org/elasticsearch/snapshots/SharedClusterSnapshotRestoreIT.java]: Sure, see 952a82bd65e885fe582f3d6961b39122f5512d93 - we can't close the index while the snapshot is running, but we can fail the shard.

**Pull request merged**